In [275]:
import pandas as pd
import re
import plotly.express as px
import numpy as np
from sklearn.preprocessing import MinMaxScaler

### RQ2.1
varying the number of parameters in the given clause

In [276]:
current_result = "RQ1_shared_scenarios_results"

In [277]:
file_path = f'res/outcome/similarities/{current_result}.csv'
raw_df = pd.read_csv(file_path, sep=';')
raw_df

,desired_scenario,cadidate_scenario,Desired Given DNF,Shared Given DNF,Penality Given,Global Similarity Given,Final Similarity Given,Desired When DNF,Shared When DNF,Penality When,Global Similarity When,Final Similarity When,Desired Then DNF,Shared Then DNF,Penality Then,Global Similarity Then,Final Similarity Then,Scenario Similarity,Action
0,Desired_1,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,takeoff == 1,takeoff == 1,0.0,1.0,1.0,height >= 30,height >= 30 AND source_distance == 0,0.090909,1.0,0.909091,0.914141,behavior_3
1,Desired_2,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,takeoff == 1,takeoff == 1,0.0,1.0,1.0,source_distance == 0,height >= 30 AND source_distance == 0,0.090909,1.0,0.909091,0.914141,behavior_3
2,Desired_3,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,takeoff == 1,takeoff == 1,0.0,1.0,1.0,a > 10,height >= 30 AND source_distance == 0,1.000000,0.0,-1.000000,0.277778,behavior_3
3,Desired_4,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,takeoff == 1,takeoff == 1,0.0,1.0,1.0,b > 10,height >= 30 AND source_distance == 0,1.000000,0.0,-1.000000,0.277778,behavior_3
4,Desired_5,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,takeoff == 1,takeoff == 1,0.0,1.0,1.0,height >= 30 AND source_distance == 0,height >= 30 AND source_distance == 0,0.000000,1.0,1.000000,0.944444,behavior_3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
460,Desired_461,candidate,battery >= 10 AND satellite_count >= 7 AND win...,battery >= 10 AND satellite_count >= 7 AND win...,0.375000,1.0,0.625000,takeoff == 1,takeoff == 1,0.0,1.0,1.0,height >= 30 AND source_distance == 0 AND a > 10,height >= 30 AND source_distance == 0,0.310345,1.0,0.689655,0.771552,behavior_3
461,Desired_462,candidate,battery >= 10 AND satellite_count >= 7 AND win...,battery >= 10 AND satellite_count >= 7 AND win...,0.375000,1.0,0.625000,takeoff == 1,takeoff == 1,0.0,1.0,1.0,height >= 30 AND source_distance == 0 AND b > 10,height >= 30 AND source_distance == 0,0.310345,1.0,0.689655,0.771552,behavior_3
462,Desired_463,candidate,battery >= 10 AND satellite_count >= 7 AND win...,battery >= 10 AND satellite_count >= 7 AND win...,0.375000,1.0,0.625000,takeoff == 1,takeoff == 1,0.0,1.0,1.0,height >= 30 AND a > 10 AND b > 10,height >= 30 AND source_distance == 0,0.655172,1.0,0.344828,0.656609,behavior_3
463,Desired_464,candidate,battery >= 10 AND satellite_count >= 7 AND win...,battery >= 10 AND satellite_count >= 7 AND win...,0.375000,1.0,0.625000,takeoff == 1,takeoff == 1,0.0,1.0,1.0,source_distance == 0 AND a > 10 AND b > 10,height >= 30 AND source_distance == 0,0.655172,1.0,0.344828,0.656609,behavior_3


In [278]:
raw_df.columns

Index(['desired_scenario', 'cadidate_scenario', 'Desired Given DNF',
       'Shared Given DNF', 'Penality Given', 'Global Similarity Given',
       'Final Similarity Given', 'Desired When DNF', 'Shared When DNF',
       'Penality When', 'Global Similarity When', 'Final Similarity When',
       'Desired Then DNF', 'Shared Then DNF', 'Penality Then',
       'Global Similarity Then', 'Final Similarity Then',
       'Scenario Similarity', 'Action'],
      dtype='object')

In [279]:
given_then_df=raw_df[['desired_scenario', 'cadidate_scenario', 'Desired Given DNF',
       'Shared Given DNF', 'Penality Given', 'Global Similarity Given',
       'Final Similarity Given','Desired Then DNF', 'Shared Then DNF', 'Penality Then',
       'Global Similarity Then', 'Final Similarity Then',
       'Scenario Similarity']]

given_then_df = given_then_df.rename(columns={'desired_scenario': 'desired_identifier',
                                               'cadidate_scenario': 'cadidate_identifier',
                                               'Desired Given DNF': 'desired_given_exp',
                                               'Shared Given DNF': 'candidate_given_exp',
                                                'Desired Then DNF': 'desired_then_exp',
                                                'Shared Then DNF': 'candidate_then_exp'})

given_then_df.head()




,desired_identifier,cadidate_identifier,desired_given_exp,candidate_given_exp,Penality Given,Global Similarity Given,Final Similarity Given,desired_then_exp,candidate_then_exp,Penality Then,Global Similarity Then,Final Similarity Then,Scenario Similarity
0,Desired_1,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,height >= 30,height >= 30 AND source_distance == 0,0.090909,1.0,0.909091,0.914141
1,Desired_2,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,source_distance == 0,height >= 30 AND source_distance == 0,0.090909,1.0,0.909091,0.914141
2,Desired_3,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,a > 10,height >= 30 AND source_distance == 0,1.000000,0.0,-1.000000,0.277778
3,Desired_4,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,b > 10,height >= 30 AND source_distance == 0,1.000000,0.0,-1.000000,0.277778
4,Desired_5,candidate,battery >= 10,battery >= 10 AND satellite_count >= 7 AND win...,0.166667,1.0,0.833333,height >= 30 AND source_distance == 0,height >= 30 AND source_distance == 0,0.000000,1.0,1.000000,0.944444


In [280]:
given_then_df.columns

Index(['desired_identifier', 'cadidate_identifier', 'desired_given_exp',
       'candidate_given_exp', 'Penality Given', 'Global Similarity Given',
       'Final Similarity Given', 'desired_then_exp', 'candidate_then_exp',
       'Penality Then', 'Global Similarity Then', 'Final Similarity Then',
       'Scenario Similarity'],
      dtype='object')

In [281]:
# Extrair variáveis do Given e Then do desejado e do candidato
def extrair_variaveis(expressao):
    variaveis = re.findall(r'(\w+)\s*(?:>=|<=|>|<|==|!=)', expressao)
    return ', '.join(variaveis)  # Opcional: juntar as variáveis em uma string separada por vírgula

# Aplicando a função para cada linha da coluna 'expressao' e armazenando na coluna 'vars'
given_then_df['desired_given_vars'] = given_then_df['desired_given_exp'].apply(extrair_variaveis)
given_then_df['candidate_given_vars'] = given_then_df['candidate_given_exp'].apply(extrair_variaveis)

given_then_df['desired_then_vars'] = given_then_df['desired_then_exp'].apply(extrair_variaveis)
given_then_df['candidate_then_vars'] = given_then_df['candidate_then_exp'].apply(extrair_variaveis)

temp=given_then_df[['desired_given_exp','desired_given_vars','candidate_given_exp','candidate_given_vars','desired_then_exp','desired_then_vars','candidate_then_exp','candidate_then_vars']]
temp

,desired_given_exp,desired_given_vars,candidate_given_exp,candidate_given_vars,desired_then_exp,desired_then_vars,candidate_then_exp,candidate_then_vars
0,battery >= 10,battery,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",height >= 30,height,height >= 30 AND source_distance == 0,"height, source_distance"
1,battery >= 10,battery,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",source_distance == 0,source_distance,height >= 30 AND source_distance == 0,"height, source_distance"
2,battery >= 10,battery,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",a > 10,a,height >= 30 AND source_distance == 0,"height, source_distance"
3,battery >= 10,battery,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",b > 10,b,height >= 30 AND source_distance == 0,"height, source_distance"
4,battery >= 10,battery,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",height >= 30 AND source_distance == 0,"height, source_distance",height >= 30 AND source_distance == 0,"height, source_distance"
...,...,...,...,...,...,...,...,...
460,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed, a, b",battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",height >= 30 AND source_distance == 0 AND a > 10,"height, source_distance, a",height >= 30 AND source_distance == 0,"height, source_distance"
461,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed, a, b",battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",height >= 30 AND source_distance == 0 AND b > 10,"height, source_distance, b",height >= 30 AND source_distance == 0,"height, source_distance"
462,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed, a, b",battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",height >= 30 AND a > 10 AND b > 10,"height, a, b",height >= 30 AND source_distance == 0,"height, source_distance"
463,battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed, a, b",battery >= 10 AND satellite_count >= 7 AND win...,"battery, satellite_count, wind_speed",source_distance == 0 AND a > 10 AND b > 10,"source_distance, a, b",height >= 30 AND source_distance == 0,"height, source_distance"


In [282]:
# Calcular a interseção entre as variáveis do Given e do Then do desejado e do candidato

def calcular_intersecao_given(row):
    # Separar as variáveis em listas
    desired_given = set(row['desired_given_vars'].split(', '))
    candidate_given = set(row['candidate_given_vars'].split(', '))
    
    # Calcular a interseção
    intersecao = desired_given & candidate_given
    return ', '.join(intersecao)  # Juntar as variáveis da interseção em uma string

given_then_df['intersection_given_vars'] = given_then_df.apply(calcular_intersecao_given, axis=1)

def calcular_intersecao_then(row):
    # Separar as variáveis em listas

    desired_then = set(row['desired_then_vars'].split(', '))
    candidate_then = set(row['candidate_then_vars'].split(', '))
    
    # Calcular a interseção
    intersecao = desired_then & candidate_then
    return ', '.join(intersecao)  # Juntar as variáveis da interseção em uma string

# Aplicando a função para cada linha e criando uma nova coluna chamada 'intersection_vars'
given_then_df['intersection_then_vars'] = given_then_df.apply(calcular_intersecao_then, axis=1)


temp=given_then_df[['desired_given_vars','candidate_given_vars','intersection_given_vars','desired_then_vars','candidate_then_vars','intersection_then_vars']]
temp

,desired_given_vars,candidate_given_vars,intersection_given_vars,desired_then_vars,candidate_then_vars,intersection_then_vars
0,battery,"battery, satellite_count, wind_speed",battery,height,"height, source_distance",height
1,battery,"battery, satellite_count, wind_speed",battery,source_distance,"height, source_distance",source_distance
2,battery,"battery, satellite_count, wind_speed",battery,a,"height, source_distance",
3,battery,"battery, satellite_count, wind_speed",battery,b,"height, source_distance",
4,battery,"battery, satellite_count, wind_speed",battery,"height, source_distance","height, source_distance","height, source_distance"
...,...,...,...,...,...,...
460,"battery, satellite_count, wind_speed, a, b","battery, satellite_count, wind_speed","satellite_count, wind_speed, battery","height, source_distance, a","height, source_distance","height, source_distance"
461,"battery, satellite_count, wind_speed, a, b","battery, satellite_count, wind_speed","satellite_count, wind_speed, battery","height, source_distance, b","height, source_distance","height, source_distance"
462,"battery, satellite_count, wind_speed, a, b","battery, satellite_count, wind_speed","satellite_count, wind_speed, battery","height, a, b","height, source_distance",height
463,"battery, satellite_count, wind_speed, a, b","battery, satellite_count, wind_speed","satellite_count, wind_speed, battery","source_distance, a, b","height, source_distance",source_distance


In [283]:
# o que tem de extra no nao referencia (given)
def given_plus(row):
    desired_given = set(row['desired_given_vars'].split(', '))
    candidate_given = set(row['candidate_given_vars'].split(', '))
    
    # Calcular a interseção
    given_plus = desired_given - candidate_given
    return ', '.join(given_plus)

# é o que tem de falta no nao referencia (given)
def given_less(row):
    desired_given = set(row['desired_given_vars'].split(', '))
    candidate_given = set(row['candidate_given_vars'].split(', '))
    
    # Calcular a interseção
    given_less = candidate_given - desired_given
    return ', '.join(given_less)

# o que tem de extra no nao referencia (then)
def then_plus(row):
    desired_then = set(row['desired_then_vars'].split(', '))
    candidate_then = set(row['candidate_then_vars'].split(', '))
    
    # Calcular a interseção
    then_plus = desired_then-candidate_then 
    return ', '.join(then_plus)

# é o que tem de falta no nao referencia (then)
def then_less(row):
    desired_then = set(row['desired_then_vars'].split(', '))
    candidate_then = set(row['candidate_then_vars'].split(', '))
    
    # Calcular a interseção
    then_less = candidate_then-desired_then 
    return ', '.join(then_less)

given_then_df['given_plus'] = given_then_df.apply(given_plus, axis=1)
given_then_df['given_less'] = given_then_df.apply(given_less, axis=1)
given_then_df['then_plus'] = given_then_df.apply(then_plus, axis=1)
given_then_df['then_less'] = given_then_df.apply(then_less, axis=1)

temp=given_then_df[["candidate_given_vars","desired_given_vars","given_plus","given_less","candidate_then_vars","desired_then_vars","then_plus","then_less"]]
temp

,candidate_given_vars,desired_given_vars,given_plus,given_less,candidate_then_vars,desired_then_vars,then_plus,then_less
0,"battery, satellite_count, wind_speed",battery,,"satellite_count, wind_speed","height, source_distance",height,,source_distance
1,"battery, satellite_count, wind_speed",battery,,"satellite_count, wind_speed","height, source_distance",source_distance,,height
2,"battery, satellite_count, wind_speed",battery,,"satellite_count, wind_speed","height, source_distance",a,a,"height, source_distance"
3,"battery, satellite_count, wind_speed",battery,,"satellite_count, wind_speed","height, source_distance",b,b,"height, source_distance"
4,"battery, satellite_count, wind_speed",battery,,"satellite_count, wind_speed","height, source_distance","height, source_distance",,
...,...,...,...,...,...,...,...,...
460,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",,"height, source_distance","height, source_distance, a",a,
461,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",,"height, source_distance","height, source_distance, b",b,
462,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",,"height, source_distance","height, a, b","a, b",source_distance
463,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",,"height, source_distance","source_distance, a, b","a, b",height


In [284]:
#Função para contar quantas variáveis existem em uma string separada por ', '
def contar_variaveis(variaveis_str):
    if variaveis_str.strip() == '':
        return 0
    return len(variaveis_str.split(', '))

# Contando o número de variáveis em 'desired_vars', 'candidate_vars' e 'intersection_vars'
given_then_df['#given_plus'] = given_then_df['given_plus'].apply(contar_variaveis)
given_then_df['#given_less'] = given_then_df['given_less'].apply(contar_variaveis)

given_then_df['#then_plus'] = given_then_df['then_plus'].apply(contar_variaveis)
given_then_df['#then_less'] = given_then_df['then_less'].apply(contar_variaveis)

temp=given_then_df[["candidate_given_vars","desired_given_vars","given_plus","#given_plus","given_less","#given_less","candidate_then_vars","desired_then_vars","then_plus","#then_plus","then_less","#then_less"]]
temp

,candidate_given_vars,desired_given_vars,given_plus,#given_plus,given_less,#given_less,candidate_then_vars,desired_then_vars,then_plus,#then_plus,then_less,#then_less
0,"battery, satellite_count, wind_speed",battery,,0,"satellite_count, wind_speed",2,"height, source_distance",height,,0,source_distance,1
1,"battery, satellite_count, wind_speed",battery,,0,"satellite_count, wind_speed",2,"height, source_distance",source_distance,,0,height,1
2,"battery, satellite_count, wind_speed",battery,,0,"satellite_count, wind_speed",2,"height, source_distance",a,a,1,"height, source_distance",2
3,"battery, satellite_count, wind_speed",battery,,0,"satellite_count, wind_speed",2,"height, source_distance",b,b,1,"height, source_distance",2
4,"battery, satellite_count, wind_speed",battery,,0,"satellite_count, wind_speed",2,"height, source_distance","height, source_distance",,0,,0
...,...,...,...,...,...,...,...,...,...,...,...,...
460,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",2,,0,"height, source_distance","height, source_distance, a",a,1,,0
461,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",2,,0,"height, source_distance","height, source_distance, b",b,1,,0
462,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",2,,0,"height, source_distance","height, a, b","a, b",2,source_distance,1
463,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed, a, b","a, b",2,,0,"height, source_distance","source_distance, a, b","a, b",2,height,1


In [285]:
given_then_df['Scenario Similarity']=given_then_df['Scenario Similarity'].round(3)

In [286]:
temp = given_then_df[['candidate_given_vars','desired_given_vars','#given_plus','#given_less','desired_then_vars','candidate_then_vars','#then_plus','#then_less','Scenario Similarity']].sort_values(by='Scenario Similarity', ascending=False)

temp

,candidate_given_vars,desired_given_vars,#given_plus,#given_less,desired_then_vars,candidate_then_vars,#then_plus,#then_less,Scenario Similarity
229,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed",0,0,"height, source_distance","height, source_distance",0,0,1.000
79,"battery, satellite_count, wind_speed","battery, satellite_count",0,1,"height, source_distance","height, source_distance",0,0,0.984
94,"battery, satellite_count, wind_speed","battery, wind_speed",0,1,"height, source_distance","height, source_distance",0,0,0.984
139,"battery, satellite_count, wind_speed","satellite_count, wind_speed",0,1,"height, source_distance","height, source_distance",0,0,0.984
225,"battery, satellite_count, wind_speed","battery, satellite_count, wind_speed",0,0,height,"height, source_distance",0,1,0.970
...,...,...,...,...,...,...,...,...,...
62,"battery, satellite_count, wind_speed",b,1,3,a,"height, source_distance",1,2,-0.333
54,"battery, satellite_count, wind_speed",a,1,3,"a, b","height, source_distance",2,2,-0.333
48,"battery, satellite_count, wind_speed",a,1,3,b,"height, source_distance",1,2,-0.333
63,"battery, satellite_count, wind_speed",b,1,3,b,"height, source_distance",1,2,-0.333


In [287]:
temp = given_then_df[['#given_plus','#given_less','#then_plus','#then_less','Scenario Similarity']].sort_values(by='Scenario Similarity', ascending=False)

temp

,#given_plus,#given_less,#then_plus,#then_less,Scenario Similarity
229,0,0,0,0,1.000
79,0,1,0,0,0.984
94,0,1,0,0,0.984
139,0,1,0,0,0.984
225,0,0,0,1,0.970
...,...,...,...,...,...
62,1,3,1,2,-0.333
54,1,3,2,2,-0.333
48,1,3,1,2,-0.333
63,1,3,1,2,-0.333


In [288]:
import plotly.express as px

# Compute the correlation matrix for the specified columns
correlation_matrix = given_then_df[['#given_plus', '#given_less', '#then_plus', '#then_less', 'Scenario Similarity']].corr()

# Generate a heatmap using Plotly
fig = px.imshow(correlation_matrix, 
                text_auto=True, 
                color_continuous_scale='RdBu_r', 
                title="Correlation Matrix")

# Show the plot
fig.show()


Com esse gráfico a gente conclui que é pior erros por falta do que por excesso. Tbm vale a destacar que o then_less é mais impactante pois ele tem menos variaveis do que o given_less, entao cada falta terá um impacto maior.

In [289]:
# given_then_df.drop_duplicates(subset=['#given_plus', '#given_less', '#then_plus', '#then_less'])[['#given_plus','#given_less','#then_plus','#then_less','Scenario Similarity']]

In [290]:
# Função para categorizar o tipo de erro com base nas regras fornecidas
def categorize_error(row):
    # Condição para "no_error"
    if row['#given_plus'] == 0 and row['#then_plus'] == 0 and row['#given_less'] == 0 and row['#then_less'] == 0:
        return 'no_error'
    # Condição para "error_alfa"
    elif row['#given_plus'] == 0 and row['#then_plus'] == 0:
        return 'error_alfa'
    # Condição para "error_beta"
    elif row['#given_less'] == 0 and row['#then_less'] == 0:
        return 'error_beta'
    # Caso contrário, é "error_alfa_beta"
    else:
        return 'error_alfa_beta'

# Aplicar a função para criar a coluna "tipo de erro"
given_then_df['tipo_de_erro'] = given_then_df.apply(categorize_error, axis=1)
temp=given_then_df[['#given_plus','#given_less','#then_plus','#then_less','Scenario Similarity','tipo_de_erro']]
temp


,#given_plus,#given_less,#then_plus,#then_less,Scenario Similarity,tipo_de_erro
0,0,2,0,1,0.914,error_alfa
1,0,2,0,1,0.914,error_alfa
2,0,2,1,2,0.278,error_alfa_beta
3,0,2,1,2,0.278,error_alfa_beta
4,0,2,0,0,0.944,error_alfa
...,...,...,...,...,...,...
460,2,0,1,0,0.772,error_beta
461,2,0,1,0,0.772,error_beta
462,2,0,2,1,0.657,error_alfa_beta
463,2,0,2,1,0.657,error_alfa_beta


In [291]:
given_then_df[["Scenario Similarity","tipo_de_erro"]].sort_values(by='Scenario Similarity', ascending=False)

,Scenario Similarity,tipo_de_erro
229,1.000,no_error
79,0.984,error_alfa
94,0.984,error_alfa
139,0.984,error_alfa
225,0.970,error_alfa
...,...,...
62,-0.333,error_alfa_beta
54,-0.333,error_alfa_beta
48,-0.333,error_alfa_beta
63,-0.333,error_alfa_beta


In [292]:
import plotly.express as px

# Criar o boxplot
fig = px.box(given_then_df, x='tipo_de_erro', y='Scenario Similarity', 
             title='Distribuição da Similaridade por Tipo de Erro',
             labels={'Scenario Similarity': 'Scenario Similarity', 'tipo_de_erro': 'Tipo de Erro'})

# Exibir o gráfico
fig.show()


In [293]:
given_then_df.columns


Index(['desired_identifier', 'cadidate_identifier', 'desired_given_exp',
       'candidate_given_exp', 'Penality Given', 'Global Similarity Given',
       'Final Similarity Given', 'desired_then_exp', 'candidate_then_exp',
       'Penality Then', 'Global Similarity Then', 'Final Similarity Then',
       'Scenario Similarity', 'desired_given_vars', 'candidate_given_vars',
       'desired_then_vars', 'candidate_then_vars', 'intersection_given_vars',
       'intersection_then_vars', 'given_plus', 'given_less', 'then_plus',
       'then_less', '#given_plus', '#given_less', '#then_plus', '#then_less',
       'tipo_de_erro'],
      dtype='object')

In [294]:
import plotly.express as px

# Criar o scatter plot
fig = px.scatter(given_then_df, x='tipo_de_erro', y='Scenario Similarity', 
                 color='tipo_de_erro', 
                 title='Distribuição da Similaridade por Tipo de Erro',
                 labels={'Scenario Similarity': 'Scenario Similarity', 'tipo_de_erro': 'Tipo de Erro'},
                 opacity=0.6,
                     hover_data={
                     'desired_given_exp': True, 
                     'candidate_given_exp': True,
                     'Penality Given': True,
                     'Global Similarity Given': True,
                    'Final Similarity Given': True,
                     'desired_then_exp': True, 
                     'candidate_then_exp': True,
                    'Penality Then': True,
                        'Global Similarity Then': True,
                        'Final Similarity Then': True,
                        '#given_plus': True,
                          '#given_less': True,
                            '#then_plus': True, '#then_less': True
                 })  # Define uma transparência para ver sobreposição de pontos

# Exibir o gráfico
fig.show()


alfa-error: quando o derivado (desejado) é errado por falta
- a,b vs a,b,c (desejado nao tem o c)

beta-error: quando o derivado (desejado) é errado por excesso
- a,b,c vs a,b (desejado tem o c de extra)

alfa-beta-erro: quando o derivado é errado por falta e excesso
- a,b,d vs a,b,c (desejado nao tem o c e tem o d de extra)

Certamente podemos dizer sem medo que temos o threshold de 0.333, pois isso significa que ou o Given ou o Then é completamente diferente do alvo

Então temos o seguinte, dependendo do tipo de erro que pode indicar um influencia negativa simples até grave. Repare que error alfa de modelagem do desejado, nao prejudica tanto encontrar o maximo global, ja o erro beta prejudica um pouco mais, mas ainda sim pouco, mas já o erro alfa-beta pode prejudicar bastante.